packages/babel-plugin-proposal-explicit-resource-management/src/index.ts

```diff
@@ -0,0 +1,227 @@
+import { declare } from "@babel/helper-plugin-utils";
+import syntaxExplicitResourceManagement from "@babel/plugin-syntax-explicit-resource-management";
+import { types as t, template, traverse, type PluginPass } from "@babel/core";
+import type { NodePath, Visitor } from "@babel/traverse";
+
+export default declare(api => {
+  // TODO: assert version 7.22.0
+  api.assertVersion(7);
+
+  const TOP_LEVEL_USING = new WeakSet();
+  const TOP_LEVEL_AWAIT_USING = new WeakSet();
+
+  function isUsingDeclaration(node: t.Node): node is t.VariableDeclaration {
+    if (!t.isVariableDeclaration(node)) return false;
+    return (
+      node.kind === "using" ||
+      node.kind === "await using" ||
+      TOP_LEVEL_USING.has(node) ||
+      TOP_LEVEL_AWAIT_USING.has(node)
+    );
+  }
+
+  const transformUsingDeclarationsVisitor: Visitor<PluginPass> = {
+    ForOfStatement(path: NodePath<t.ForOfStatement>) {
+      const { left } = path.node;
+      if (!isUsingDeclaration(left)) return;
+
+      const { id } = left.declarations[0];
+      const tmpId = path.scope.generateUidIdentifierBasedOnNode(id);
+      left.declarations[0].id = tmpId;
+      left.kind = "const";
+
+      path.ensureBlock();
+      path.node.body.body.unshift(
+        t.variableDeclaration("using", [
+          t.variableDeclarator(id, t.cloneNode(tmpId)),
+        ]),
+      );
+    },
+    "BlockStatement|StaticBlock"(
+      path: NodePath<t.BlockStatement | t.StaticBlock>,
+      state,
+    ) {
+      let stackId: t.Identifier | null = null;
+      let needsAwait = false;
+
+      for (const node of path.node.body) {
+        if (!isUsingDeclaration(node)) continue;
+        stackId ??= path.scope.generateUidIdentifier("stack");
+        const isAwaitUsing =
+          node.kind === "await using" || TOP_LEVEL_AWAIT_USING.has(node);
+        needsAwait ||= isAwaitUsing;
+
+        if (
+          !TOP_LEVEL_AWAIT_USING.delete(node) &&
+          !TOP_LEVEL_USING.delete(node)
+        ) {
+          node.kind = "const";
+        }
+        node.declarations.forEach(decl => {
+          const args = [t.cloneNode(stackId), decl.init];
+          if (isAwaitUsing) args.push(t.booleanLiteral(true));
+          decl.init = t.callExpression(state.addHelper("using"), args);
+        });
+      }
+      if (!stackId) return;
+
+      const errorId = path.scope.generateUidIdentifier("error");
+      const hasErrorId = path.scope.generateUidIdentifier("hasError");
+
+      let disposeCall: t.Expression = t.callExpression(
+        state.addHelper("dispose"),
+        [
+          t.cloneNode(stackId),
+          t.cloneNode(errorId),
+          t.cloneNode(hasErrorId),
+          // Pass SuppressedError so that it can be used with "pure"
+          // polyfills that do not compile the contents of runtime
+          // helpers.
+          template.expression.ast`
+            typeof SuppressedError !== "undefined" && SuppressedError
```

Comment:
Can we make it a helper? The minifier can't do much thing about global variables.

Reply:
The reason I put this here and not inline code is so that if you ran `babel-plugin-polyfill-corejs3` in `"usage-pure"` mode it has a change of injecting the polyfill. If the helper relied on a global, it wouldn't work with "pure" polyfills.

Reply:
Personally I prefer to create a variable `_SuppressedError` at the top level and then always reference it.
By the way, will we merge it soon? I want to try it https://github.com/babel/babel/pull/15633#issuecomment-1556634075 :)

Reply:
I just realized that thanks to https://github.com/babel/babel/pull/15531, when using `"usage-pure"` users can get imports to the polyfilled version of the helper. We don't need to pass `SuppressedError` from the user code.